hazelcast-client-new/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -0,0 +1,203 @@
+    public ClientDelegatingFuture(ClientInvocationFuture clientInvocationFuture,
+                                  SerializationService serializationService, FutureValue defaultValue) {
+        this.future = clientInvocationFuture;
+        this.serializationService = serializationService;
+        this.defaultValue = defaultValue;
+    }
+
+    public ClientDelegatingFuture(ClientInvocationFuture clientInvocationFuture,
+                                  SerializationService serializationService) {
+        this.future = clientInvocationFuture;
+        this.serializationService = serializationService;
+        this.defaultValue = null;
+    }
+
+    public void andThenInternal(final ExecutionCallback<Data> callback) {
+        future.andThenInternal(new DelegatingExecutionCallback<Data>(callback, false));
+    }
+
+    @Override
+    public void andThen(final ExecutionCallback<FutureValue> callback) {
+        future.andThen(new DelegatingExecutionCallback<FutureValue>(callback, true));
+    }
+
+    @Override
+    public void andThen(final ExecutionCallback<FutureValue> callback, Executor executor) {
+        future.andThen(new DelegatingExecutionCallback<FutureValue>(callback, true), executor);
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        done = true;
+        return false;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return false;
+    }
+
+    @Override
+    public final boolean isDone() {
+        return done ? done : future.isDone();
+    }
+
+    @Override
+    public FutureValue get() throws InterruptedException, ExecutionException {
+        try {
+            return get(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            return ExceptionUtil.sneakyThrow(e);
+        }
+    }
+
+    @Override
+    public FutureValue get(long timeout, TimeUnit unit) throws InterruptedException,
+            ExecutionException, TimeoutException {
+        if (!done) {
+            synchronized (mutex) {
+                if (!done) {
+                    try {
+                        value = resolveMessageToValue(future.get(timeout, unit));
+                    } catch (InterruptedException e) {
+                        error = e;
+                    } catch (ExecutionException e) {
+                        error = e;
+                    }
+                    done = true;
+                }
+            }
+        }
+        if (error != null) {
+            if (error instanceof ExecutionException) {
+                throw (ExecutionException) error;
+            }
+            if (error instanceof CancellationException) {
+                throw (CancellationException) error;
+            }
+            if (error instanceof InterruptedException) {
+                throw (InterruptedException) error;
+            }
+            // should not happen!
+            throw new ExecutionException(error);
+        }
+        return getResult();
```

Comment:
Yes, it will be. I did not cache the deserialized object. I could not change this one because I need to pass data to an internal call back(one expecting data as response) . And to do that I ended up with implementing an unnecessary deserialize and serialize back to data.